pkg/apis/apis/v1alpha1/types.go

```diff
@@ -0,0 +1,232 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	conditionsv1alpha1 "github.com/kcp-dev/kcp/third_party/conditions/apis/conditions/v1alpha1"
+)
+
+// ServiceBinding projects APIs and backing implementations into a Workspace.
+//
+// +crd
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:subresource:status
+// +kubebuilder:resource:scope=Cluster,categories=kcp
+type ServiceBinding struct {
+	metav1.TypeMeta `json:",inline"`
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec holds the desired state.
+	// +optional
+	Spec ServiceBindingSpec `json:"spec,omitempty"`
+
+	// Status communicates the observed state.
+	// +optional
+	Status ServiceBindingStatus `json:"status,omitempty"`
+}
+
+// ServiceBindingSpec records the APIs and implementations that are to be bound.
+type ServiceBindingSpec struct {
+	// Services describe the APIs and implementations that are to be bound.
+	Services []ServiceSpec `json:"services,omitempty"`
+}
+
+// ServiceSpec describes one API and backing implementation.
+type ServiceSpec struct {
+	Native   *NativeServiceSpec   `json:"native,omitempty"`
+	Local    *LocalServiceSpec    `json:"local,omitempty"`
+	External *ExternalServiceSpec `json:"external,omitempty"`
+}
+
+// NativeServiceSpec describes an API and backing implementation that are provided by an actor on kcp.
+type NativeServiceSpec struct {
+	// Name of the ServiceExport that describes the API for this NativeServiceSpec.
+	//
+	// +kubebuilder:validation:Required
+	Name string `json:"name"`
+
+	// Workspace of the ServiceExport that describes the API for this NativeServiceSpec.
+	//
+	// +kubebuilder:validation:Required
+	Workspace string `json:"workspace"`
+}
+
+// LocalServiceSpec describes an API and backing implementation that are provided by an actor in the
+// same Workspace as the ServiceBinding.
+type LocalServiceSpec struct {
+	// Name of the ServiceExport that describes the API for this LocalServiceSpec.
+	//
+	// +kubebuilder:validation:Required
+	Name string `json:"name"`
+}
+
+// ExternalServiceSpec describes an API and backing implementation that are provided by a third-party actor.
+type ExternalServiceSpec struct {
+	// URL at which this external service may be reached.
+	//
+	// +kubebuilder:validation:Format:uri
+	// +kubebuilder:validation:Required
+	URL string `json:"url"`
+}
+
+// ServiceBindingStatus records which schemas are bound.
+type ServiceBindingStatus struct {
```

Comment:
Removed the suggested observedGeneration here since we have them in conditions for each bound service.